rllib/utils/nested_dict.py

```diff
@@ -0,0 +1,305 @@
+        key value tuples where the values can themselves be recursively the values
+        that a nested dict can take.
+
+    Args:
+        x: a representation of a nested dict: it can be an iterable of `SeqStrType`
+        to values. e.g. `[(("a", "b") , 1), ("b", 2)]` or a mapping of flattened
+        keys to values. e.g. `{("a", "b"): 1, ("b",): 2}` or any nested mapping,
+        e.g. `{"a": {"b": 1}, "b": 2}`.
+
+    Example:
+        Basic usage:
+            >>> foo_dict = NestedDict()
+            >>> # Setting elements, possibly nested:
+            >>> foo_dict['a'] = 100         # foo_dict = {'a': 100}
+            >>> foo_dict['b', 'c'] = 200    # foo_dict = {'a': 100, 'b': {'c': 200}}
+            >>> foo_dict['b', 'd'] = 300    # foo_dict = {'a': 100,
+            >>>                             #             'b': {'c': 200, 'd': 300}}
+            >>> foo_dict['b', 'e'] = {}     # foo_dict = {'a': 100,
+            >>>                             #            'b': {'c': 200, 'd': 300}}
+            >>> # Getting elements, possibly nested:
+            >>> print(foo_dict['b', 'c'])   # 200
+            >>> print(foo_dict['b']) # IndexError("Use get for partial indexing.")
+            >>> print(foo_dict.get('b'))    # {'c': 200, 'd': 300}
+            >>> print(foo_dict) # {'a': 100, 'b': {'c': 200, 'd': 300}}
+            >>> # Converting to a dict:
+            >>> foo_dict.asdict()  # {'a': 100, 'b': {'c': 200, 'd': 300}}
+            >>> # len function:
+            >>> print(len(foo_dict))  # 3
+            >>> # Iterating:
+            >>> foo_dict.keys()  # dict_keys(['a', ('b', 'c'), ('b', 'd')])
+            >>> foo_dict.items() # dict_items([('a', 100), (('b', 'c'), 200), (('b',
+            'd'), 300)])
+            >>> foo_dict.shallow_keys()  # dict_keys(['a', 'b'])
+        Filter:
+            >>> dict1 = NestedDict([
+                (('foo', 'a'), 10), (('foo', 'b'), 11),
+                (('bar', 'c'), 11), (('bar', 'a'), 110)])
+            >>> dict2 = NestedDict([('foo', NestedDict(dict(a=11)))])
+            >>> dict3 = NestedDict([('foo', NestedDict(dict(a=100))),
+                                    ('bar', NestedDict(dict(d=11)))])
+            >>> dict4 = NestedDict([('foo', NestedDict(dict(a=100))),
+                                    ('bar', NestedDict(dict(c=11)))])
+            >>> dict1.filter(dict2).asdict()   # {'foo': {'a': 10}}
+            >>> dict1.filter(dict4).asdict()   # {'bar': {'c': 11}, 'foo': {'a': 10}}
+            >>> dict1.filter(dict3).asdict()   # KeyError - ('bar', 'd') not in dict1
+    """
+
+    def __init__(
+        self,
+        x: Optional[NestedDictInputType] = None,
+    ):
+        # shallow dict
+        self._data = dict()  # type: Dict[str, Union[T, NestedDict[T]]]
+        x = x or {}
+        if isinstance(x, Mapping):
+            for k, v in x.items():
+                self[k] = v
```

Comment:
Looking at the code, a lot of things are more complicated because a value can be a regular Mapping.
like say in the middle of this NestedDict, you all of a sudden get a big regular dict.
then this subtree doesn't support indexing with ['a', 'b', 'c'] syntax.
for the specific use case of NestedDict, can we enforce that everything must be NestedDict?
basically, we do ``if isinstance(v, Mapping): self[k] = NestedDict(v) else self[k] = v`` here.
then ``NestDict(some_dict)`` will become an actual full NestedDict.

Reply:
We already do. `__setitem__` converts any Mapping to NestedDict by default. Now that you mentioned it, I added an if statement in `__init__` to avoid doing it twice.

Reply:
There is also a unittest exactly for this kinda behavior you mentioned.